circuit-lib/src/elgamal-babyjub/elgamal.ts

```diff
@@ -0,0 +1,444 @@
+
+  return privKey;
+}
+
+/**
+ * @return A BabyJub-compatible private key.
+ */
+const genPrivKey = (): PrivKey => {
+  return genRandomBabyJubValue();
+};
+
+/**
+ * @return A BabyJub-compatible salt.
+ */
+const genRandomSalt = (): PrivKey => {
+  return genRandomBabyJubValue();
+};
+
+/**
+ * @param privKey A private key generated using genPrivKey()
+ * @return A public key associated with the private key
+ */
+function genPubKey(privKey: PrivKey): PubKey {
+  // Check whether privKey is a field element
+  privKey = BigInt(privKey.toString());
+  assert(privKey < SNARK_FIELD_SIZE);
+  return ElGamalUtils.prv2pub(ElGamalUtils.bigInt2Buffer(privKey));
+}
+
+/**
+ *
+ * @returns A keypair containing a BabyJub-compatible private key with its associated public key.
+ */
+function genKeypair(): Keypair {
+  const privKey = genPrivKey();
+  const pubKey = genPubKey(privKey);
+
+  const Keypair: Keypair = { privKey, pubKey };
+
+  return Keypair;
+}
+
+/**
+ *
+ * @returns A random extended point on the Baby Jubjub curve.
+ */
+function genRandomPoint(): BabyJubExtPoint {
+  const salt = genRandomBabyJubValue();
+  return genPubKey(salt);
+}
+
+/**
+ * Encrypts a plaintext such that only the owner of the specified public key
+ * may decrypt it.
+ * @param pubKey The recepient's public key
+ * @param encodedMessage A plaintext encoded as a BabyJub curve point (optional)
+ * @param randomVal A random value y used along with the private key to generate the ciphertext (optional)
+ */
+function encrypt(
+  pubKey: PubKey,
+  encodedMessage?: BabyJubExtPoint,
+  randomVal?: bigint,
+) {
+  const message = encodedMessage ?? genRandomPoint();
+
+  // The sender chooses a secret key as a nonce
+  const nonce: bigint =
+    randomVal ?? ElGamalUtils.formatPrivKeyForBabyJub(genRandomSalt());
+
+  // The sender calculates an ephemeral key => [nonce].Base
+  const ephemeral_key = babyJub.BASE.multiply(nonce);
+  const masking_key = pubKey.multiply(nonce);
+  let encrypted_message: BabyJubExtPoint;
+  // The sender encrypts the encodedMessage
+  if (pubKey.assertValidity && !pubKey.equals(babyJub.ZERO)) {
+    encrypted_message = message.add(masking_key);
+  } else throw new Error("Invalid Public Key!");
+
+  return { message, ephemeral_key, encrypted_message, nonce };
+}
+
+/**
+ * Decrypts a ciphertext using a private key.
+ * @param privKey The private key
+ * @param ciphertext The ciphertext to decrypt
+ */
+function decrypt(
+  privKey: PrivKey,
+  ephemeral_key: BabyJubExtPoint,
+  encrypted_message: BabyJubExtPoint,
+): BabyJubExtPoint {
+  // The receiver decrypts the message => encryptedMessage - [privKey].ephemeralKey
+  const masking_key = ephemeral_key.multiply(
+    ElGamalUtils.formatPrivKeyForBabyJub(privKey),
+  );
+  const decrypted_message = encrypted_message.add(masking_key.negate());
+
+  return decrypted_message;
+}
+
+/**
+ *  ElGamal Encryption function with specified inputs for testing purposes
+ */
+function encrypt_demo(
```

Comment:
please move this function to the test or a test-utils folder in tests/test-utils/elgamal